src/cpu/x64/matmul/amx_blocking_heuristics.hpp

```diff
@@ -0,0 +1,261 @@
+            matmul_amx_blocking_params_macro_t &best_blocking);
+
+protected:
+    float calculate_blocking_scores() override;
+
+private:
+    static const dim_t min_m_dim = 64;
+    static const dim_t min_k_dim = 256;
+    static const dim_t min_n_dim = 64;
+    static const dim_t k_threshold_write_bound_layer = 256;
+    static const dim_t min_n_dim_write_bound_layer = 256;
+    dim_t n_decomposition = 32;
+    dim_t m_decomposition = 32;
+    size_t gemm_dt_sz;
+    dim_t m_per_thread, k_per_thread, n_per_thread, b_per_thread;
+    bool need_prefetch;
+    bool is_horizontal;
+
+    size_t m_tmul, n_tmul, k_tmul;
+    bool set_blocking_parameters();
+    bool is_horizontal_selected(bool horizontal_not_possible,
+            bool vertical_not_possible, size_t best_m_v, size_t best_k_v,
+            size_t k_blk_v) const;
+    void set_tmul_sizes();
+    void set_decomposition();
+    size_t l2_matrix_usage(size_t k_chunk_size, size_t m_or_n_blk, size_t k_blk,
+            bool is_horizontal);
+    size_t l2_matrix_and_c_usage(size_t k_chunk_size, size_t m_or_n_blk,
+            size_t k_blk, bool is_horizontal);
+    void set_core_divs(int nthr_b, int nthr_m, int nthr_k, int nthr_n);
+    int bw(size_t m_blk, size_t k_chunk_size, size_t k_blk, size_t n_blk,
+            bool is_horizontal);
+    int compute(size_t m_blk, size_t k_chunk_size, size_t k_blk,
+            size_t n_blk) const;
+    float ratio(size_t m_blk, size_t k_chunk_size, size_t k_blk, size_t n_blk,
+            bool is_horizontal);
+    std::set<dim_t> blk_candidates(dim_t dim_per_thread, dim_t decomposition);
+    float evaluate_single_core_blocking(size_t k_chunk_size, size_t m_or_n_blk,
+            size_t k_blk, bool is_horizontal);
+    dim_t calc_k_blk(size_t l1_dim);
+    bool divs_are_acceptable();
+    bool operator>(matmul_amx_blocking_params_macro_t &other);
+};
+
+class matmul_amx_blocking_params_micro_t : public matmul_amx_blocking_params_t {
+public:
+    matmul_amx_blocking_params_micro_t() = default;
+
+    matmul_amx_blocking_params_micro_t(const brgemm_matmul_conf_t &bgmmc)
+        : matmul_amx_blocking_params_t(bgmmc) {}
+
+    void set_blocking_parameters(int nthr_k, int n_blk, int n_chunk_size,
+            int m_blk, int m_chunk_size);
+
+    static void find_best_blocking(const brgemm_matmul_conf_t &bgmmc,
+            const brgemm_matmul_conf_utils_t &bm_conf_utils,
+            matmul_amx_blocking_params_t &best_blocking);
+
+protected:
+    float calculate_blocking_scores() override;
+
+private:
+    float get_thread_balance_scores();
+    void update_k_blocking_dependent_params();
+    size_t calculate_chunk_memory_size();
+    float get_copied_data_reusage_scores();
+    float get_L2_utilization_scores() const;
+};
+
+class bw_map_t {
+public:
+    bw_map_t() {}
+
+    float get_bw(int x) { return linear_interpolation(multicore_bw, x); }
+
+    //All the following bandwidth measurements were taken on an
```

Comment:
It would be nice to clean up comments across the patch and have all them contain a starting space, at least. Capitilized letters could be a nice bonus.